crates/sui-framework/sources/safe.move

```diff
@@ -0,0 +1,178 @@
+    const INVALID_OWNER_CAPABILITY: u64 = 1;
+    const TRANSFER_CAPABILITY_REVOKED: u64 = 2;
+    const OVERDRAWN: u64 = 3;
+
+
+    ///
+    /// Allows any holder of a capability to transfer a fixed amount of assets from the safe.
+    /// Useful in situations like an NFT marketplace where you wish to buy the NFTs at a specific price.
+    /// 
+    /// @ownership: Shared
+    /// 
+    struct Safe<phantom T> has key {
+        id: UID,
+        balance: Balance<T>,
+        permissions_bitmap: BitVector,
+        next_nonce: u64,
+    }
+
+    struct OwnerCapability<phantom T> has key, store {
+        id: UID,
+        safe_id: ID,
+    }
+
+    ///
+    /// Allows the owner of the capability to take `amount` of coins from the box.
+    ///
+    /// @ownership: Owned
+    ///
+    struct TransferCapability<phantom T> has store, key {
+        id: UID,
+        safe_id: ID,
+        nonce: u64,
+        // The amount that the user is able to transfer.
+        amount: u64,
+    }
+
+    //////////////////////////////////////////////////////
+    /// HELPER FUNCTIONS
+    //////////////////////////////////////////////////////
+    
+    /// Check that the capability has not yet been revoked by the owner.
+    fun check_capability_validity<T>(safe: &Safe<T>, capability: &TransferCapability<T>) {
+        // Check that the ids match
+        assert!(object::id(safe) == capability.safe_id, INVALID_TRANSFER_CAPABILITY);
+        // Check that it has not been cancelled
+        assert!(bit_vector::is_index_set(&safe.permissions_bitmap, capability.nonce), TRANSFER_CAPABILITY_REVOKED);
+    }
+
+    fun check_owner_capability_validity<T>(safe: &Safe<T>, capability: &OwnerCapability<T>) {
+        assert!(object::id(safe) == capability.safe_id, INVALID_OWNER_CAPABILITY);
+    }
+
+    /// Helper function to create a capability.
+    fun create_capability_<T>(safe: &mut Safe<T>, withdraw_amount: u64, ctx: &mut TxContext): TransferCapability<T> {
+        let capability = TransferCapability {
+            id: object::new(ctx),
+            safe_id: object::uid_to_inner(&safe.id),
+            nonce: safe.next_nonce,
+            amount: withdraw_amount,
+        };
+
+        bit_vector::set(&mut safe.permissions_bitmap, safe.next_nonce);
+
+        safe.next_nonce = safe.next_nonce + 1;
+        capability
+    }
+
+    //////////////////////////////////////////////////////
+    /// PUBLIC FUNCTIONS
+    //////////////////////////////////////////////////////
+    
+    public fun balance<T>(safe: &Safe<T>): &Balance<T> {
+        &safe.balance
+    }
+
+    /// Wrap a coin around a safe.
+    /// a trusted party (or smart contract) to transfer the object out.
+    public fun create_<T>(balance: Balance<T>, ctx: &mut TxContext): OwnerCapability<T> {
+        let safe = Safe {
+            id: object::new(ctx),
+            balance,
+            next_nonce: 0,
+            permissions_bitmap: bit_vector::new(1000),
```

Comment:
The other idea is to revert to the previous bitmap implementation that I had, it adds on a u64 every time that it needs to extend the length of the bitmap, and this means that on average, each capability adds on 1 bit. This of course does not have the benefit of being able to clean up space when revoking an ID. We would be able to support a revoke_all API by simply looping until next_nonce, and disabling all the capabilities. But of course, this does not have the benefits of the lookup.

Reply:
Leaving this up to you!

Reply:
Your other bitmap would indeed be more space-efficient than the bitvector library (sorry for suggesting you switch to that and then criticizing the choice lol, I forgot how inefficient that library is...).

I would vote for going with the `vector<ID>` approach mostly because it's (a) very simple and (b) not ever-growing--I think both the vector and the bitvector approach can comfortably support hundreds of simultaneous capabilities, which seems good enough. A `Safe` that needs thousands would probably want to use one of the forthcoming maps or something else.